### ISSUE TITLE
sql: warn users that sequences are slower than using UUID

### ISSUE THREAD

**Issue by awoods187** [cockroachdb/cockroach#41258]
In our docs we write:

> Using a sequence is slower than auto-generating unique IDs with the gen_random_uuid(), uuid_v4() or unique_rowid() built-in functions. Incrementing a sequence requires a write to persistent storage, whereas auto-generating a unique ID does not. Therefore, use auto-generated unique IDs unless an incremental sequence is preferred or required.

https://www.cockroachlabs.com/docs/stable/create-sequence.html#considerations

This is great information but many users will not read the docs in advance or understand why usign sequences/sequential ids are slower in CRDB. They also may not be aware that better alternatives exist. We should return a warning in the SQL shell when users create these types of patterns:

```SQL
CREATE TABLE customer_list (
    id INT PRIMARY KEY DEFAULT nextval('customer_seq'),
    customer string,
    address string
  );
CREATE TABLE

Time: 123.133ms
*Warning: Using a sequence is slower than auto-generating unique IDs with the gen_random_uuid() or uuid_v4()  built-in functions. Click here to learn more
```

Other places to apply this warning are when using anything that would create sequential ids such as unique_rowid() (even though its faster than sequences it can still be susceptible).

Epic CRDB-8948

**Comment by rafiss** [comment]
We could also warn on the serial_normalization setting being changed

**Comment by knz** [comment]
I think this needs a regular pg notification, like we already do for other things -- the phrase "warning in the sql shell" is a bit too vague otherwise.

**Comment by rafiss** [comment]
Agree, that's the approach we'll take for this. I think the issue was filed before we added support for PG notices.

**Comment by rafiss** [comment]
We may also want to direct users to cached sequences https://github.com/cockroachdb/cockroach/issues/51259 -- though if we want to just pick one piece of advice then I think it should be to use `gen_random_uuid`

**Comment by iAziz786** [comment]
I would like to work on this.

**Comment by rafiss** [comment]
thanks @iAziz786 ! to understand how notices work, take a look at a few examples:

https://github.com/cockroachdb/cockroach/blob/2efbbb25456c21d3878f23b7adf90dd2bc2a047c/pkg/sql/alter_column_type.go#L141-L143

https://github.com/cockroachdb/cockroach/blob/2efbbb25456c21d3878f23b7adf90dd2bc2a047c/pkg/sql/create_type.go#L97-L99

For this issue, we want a notice in create_table.go. In startExec, check if the statement contains a column that is a PRIMARY KEY with a type of `SERIAL` or a default value of `nextval(...)`. The notice should contain a link to this docs page: https://www.cockroachlabs.com/docs/v21.1/serial.html. The message can say "using sequential values in a primary key does not perform as well as using random UUIDs. See https://www.cockroachlabs.com/docs/v21.1/serial.html"

To create the url, use `func URL(pageName string)` from the `docs` package.

**Comment by iAziz786** [comment]
@rafiss Thanks for the clarification. I have a few questions:

1. If we choose to change the `startExec` file, do I also need to change any test case https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/create_table?

2. Or if I decide to add the default value in `nextval(...)` do I need to change this function? https://github.com/cockroachdb/cockroach/blob/73154b35a3f679861a102b7cfcbeaa9bfc1aa10f/pkg/sql/sessiondata/sequence_cache.go#L42
3. After some digging I found that the column name I will get from [`IndexElem`](https://github.com/cockroachdb/cockroach/blob/73154b35a3f679861a102b7cfcbeaa9bfc1aa10f/pkg/sql/sem/tree/create.go#L117) but can you help me to find where can I get the type of the column?

Thanks :)

**Comment by rafiss** [comment]
1. Yes, see this example for how it can be tested https://github.com/cockroachdb/cockroach/pull/63326
2. You don't need to change `sequence_cache.go` -- for this issue, the notice is only sent when using a `CREATE TABLE` statement.
3. You will need to loop over `n.n.Defs`; each element is a `TableDef`. In the loop you should have a type switch on the `def` with cases for `*tree.UniqueConstraintTableDef` and `*tree.ColumnTableDef`. For unique constraints, check if the def is `def.PrimaryKey` and keep track of the columns in the primary key (`def.Columns`). For the `ColumnTableDef` case, you can use `def.PrimaryKey.IsPrimaryKey`, `def.IsSerial`, and also check if `def.DefaultExpr.Expr.(type)` is a `FuncExpr` named `nextval`. (This logic will have a lot of cases to check, so let me know if you want me to write out more specific code.)

**Comment by iAziz786** [comment]
Thanks, let me try it once. Just one small help, does this command will run the test cases for `create_table`?

```
make test PKG=./pkg/sql/logictest TESTS=TestLogic
```

Or do I need to run a different command in order to do so?

**Comment by rafiss** [comment]
That command will run all of the logic tests -- you will find it faster to use `make testlogic FILES='create_table'`

Good luck!

**Comment by iAziz786** [comment]
Reaching towards the solution, can you please help me with these new questions:

1. The `*tree.ColumnTableDef` looks like this, does calling `String()` provides me the exact name of the function?
```go
	case *tree.ColumnTableDef: {
		if d.PrimaryKey.IsPrimaryKey && d.IsSerial {
			switch f := d.DefaultExpr.Expr.(type) {
			case *tree.FuncExpr:
				f.String() // will this provide me the name of the function?
			}
		}
	}
```
2. In `*tree.UniqueConstraintTableDef`, can you please elaborate on what it means by keeping track of the column? This is what I thought. I'm pretty sure it's wrong.
```go
	case *tree.UniqueConstraintTableDef: {
		if d.PrimaryKey {
			for _, indexElem := range d.Columns {
				if indexElem.Column.String() == "PRIMARY" {

				}
				if indexElem.Column.String() == "KEY" {

				}
			}
		}
	}
```
3. In the `create_table` file of the test cases I have noticed there is `query T noticetrace` for the queries which produce notices, what would be that in my case if my statement looks like this:
```
statement warn_slow_sequence

CREATE TABLE customer_list (
  id INT PRIMARY KEY DEFAULT nextval('customer_seq'),
  customer string,
  address string
);

query T noticetrace
???
----
NOTICE: using sequential values in a primary key does not perform as well as using random UUIDs.
```

##### For Tests
I have noticed something strange while running the test cases. Even if I only provide the subtest (`make testlogic FILES='create_table' SUBTESTS='warn_slow_sequence'`), it is like running forever. Might be issue at my end.

Regards

**Comment by rafiss** [comment]
This warning will be included in v22.1
